### PR TITLE
chore(rust): restructure Rust plugins as independent crates

### LINKED ISSUE #2730: [RUST] Plugin Architecture Options
**Problem:**
Currently `pii_filter` is not a subcrate and has plugin has PyO3 dependencies and macros embedded directly. This makes adding new plugins harder and couples plugin logic to Python bindings.

---

## Context

As we expand support for both Rust and Python implementations, it would be good to align on **testing strategy** and **plugin architecture** to reduce duplication and long-term maintenance cost.

This issue proposes:

* Using **Python for integration testing** shared across Rust and Python implementations.
* Evaluating **three possible plugin architecture approaches**, with trade-offs for each.

---

## Architecture

Below are three possible architectural approaches for plugins, listed from most Rust-centric to most Python-centric.

---

### Option 1: Rust API Built on Top of the Python API

**Description**
Implement a complete Rust API that wraps or mirrors the Python API. Rust plugins would be written in pure Rust, without `pyo3` or `maturin`.

**Pros**

* Pure Rust plugin development experience
* No Python bindings required for plugins
* Clean separation between core logic and Python runtime

**Cons**

* No ability to ship a `pip` package
* Rust API becomes a public contract that must be maintained indefinitely
* External users may depend on this Rust API (maintenance burden)

> This can be seen as either a feature or a liability, depending on how much we want to support external Rust consumers.

---

### Option 2: Fully Independent Plugin Crates

**Description**
Each plugin lives in its own crate and exposes its own `#[pyfunction]` / `#[pymodule]` for in-process usage, typically packaged and installed via `maturin`.
Optionally, a plugin may also expose a **gRPC or HTTP interface** instead of (or in addition to) Python bindings, if the implementer prefers an out-of-process model. Also helper crates can be developed gRPC crate for example.

**Pros**

* Clear ownership and isolation per plugin
* Straightforward Python integration for in-process use
* Natural fit for distribution via `pip`
* Plugin authors can choose between in-process (PyO3) or out-of-process (gRPC/HTTP) execution

**Cons**

* Each plugin must implement and maintain its own bindings or API surface
* Slightly more boilerplate per plugin
* Mixed execution models may increase overall system complexity

> Binding complexity is relatively low based on initial experiments, but repetition remains.

---

### Option 3: Hybrid Workspace with a Dedicated Adapter Crate

**Description**

* Plugins are written as **pure Rust subcrates** in a workspace
* One or more **adapter crates** provide integration layers (e.g. PyO3 adapter, gRPC/HTTP adapter)
* Adapters expose a unified interface while plugins remain runtime-agnostic

**Pros**

* Plugins remain clean, pure Rust
* Centralized and consistent integration layers
* Supports multiple adapters (in-process Python, gRPC/HTTP, etc.) without modifying plugins
* Avoids duplication while keeping plugin crates simple

**Cons**

* Adapter crates become critical integration points
* Slightly more complex workspace and dependency structure
* Requires careful API design between plugins and adapters

---

### Option 4: Only use gRPC or HTTP

**Description**
Expose plugin functionality over a gRPC or HTTP API, potentially running in a separate process with its own resource management.

**Pros**

* Clear language and runtime separation
* Plugins can run in an isolated process with dedicated resources
* Easier to integrate with non-Python / non-Rust consumers
* Failure isolation (plugin crashes don’t bring down the host)

**Cons**

* Added latency compared to in-process calls
* More complex setup and deployment (network API, service lifecycle)
* Increased operational and debugging complexity
* Harder local development experience

> Performance impact would need to be evaluated; in some cases isolation and resource control may outweigh the overhead.

---

Currently leaning toward options 2 or 3 and could have different adapter crates one for pyo3 and one for the Grpc giving flexibility.

---

### Proposal

Use **Python-based integration tests** that are shared between Rust and Python implementations (when both are present).

### Rationale

* Integration tests are focused on behavior, not implementation details.
* The code will not be run from native Rust directly anyway.
* A single test suite avoids duplicated effort and reduces drift between implementations.
* Python is already well-suited for higher-level integration testing and orchestration.

### Benefits

* One source of truth for integration behavior
* Easier to add new test cases once
* Consistent validation across language boundaries

### Potential Drawbacks

* Rust-only contributors may need minimal Python familiarity
* Debugging failures may be slightly less Rust-centric

---

From the Python side, there is also some repetition that could likely be removed. This is something we can revisit and simplify regardless of which architectural option we choose.

### LINKED ISSUE #2906: [CHORE]: Fix CI rust env and apply same corrections to `pii_filter` plugin
**Description:**

Current CI there is duplication (see context in PR #2785 comment: https://github.com/IBM/mcp-context-forge/pull/2785#issuecomment-3895623330).

Issues:
- duplicated testing

**Tasks:**
- [x] Enable Rust env in main pytest CI, and make sure the parameterized rust test are never ignored in pytest CI (enforce it)
- [x] Remove the original pytest specialized / separate CI for Rust plugins (and related Rust env setup in main pytest)
- [x] for `PII_filter` merge the test with rust in the name with the python plugin tests to one parameterized test suite (integration, unit, differential),  
- [x] Refactor tests to use parameterized approach with `use_rust` flag (run both Python and Rust implementations from main pytest; skip Rust-specific cases when inapplicable), parameterization can skip internal unit tests function not present in rust
- [x] Ensure both plugins pass CI (unit tests + lint + type check) in the new architecture
- [x] Verify no breaking changes in behavior
- [x] Remove `develop` dependency from `test-python` Makefile target to prevent duplicate Rust test execution in python-integration CI job

**Acceptance criteria:**
- CI run parameterized tests (rust/python)
- `pii_filter` special tests are merged and conformed with python tests
- No new test failures introduced
- Separate Rust-only pytest CI removed; all tests consolidated under main pytest with parameterization

**Additional context**  
Reference comment: https://github.com/IBM/mcp-context-forge/pull/2785#issuecomment-3895623330

### PR DESCRIPTION
**Pull request by lucarlig**
Closes #2730
Closes #2906

---

## 🚀 Summary

Restructures Rust plugins as independent crates with proper workspace architecture and consolidates CI testing to eliminate duplication. The PII filter plugin testing is migrated to the new structure as the reference implementation.

**Key changes:**
- Rust plugins now live as independent subcrates in `plugins_rust/` workspace
- Each plugin has its own `Cargo.toml`, `Makefile`, and isolated build system
- Unified parameterized testing approach for both Python and Rust implementations
- Removed duplicate Rust-specific CI workflow in favor of consolidated pytest CI
- PII filter tests merged into single parameterized suite

---

## 🧪 Checks

- [x] `make lint` passes
- [x] `make test` passes
- [x] CHANGELOG updated (if user-facing)

---

### Architecture Decision

This PR implements **Option 2: Fully Independent Plugin Crates** from issue #2730:

- Each Rust plugin is a standalone crate with its own PyO3 bindings
- Plugins can be built, tested, and distributed independently
- Clear ownership and isolation per plugin
- Natural fit for `pip` distribution via `maturin`

### Testing Strategy

Implements Python-based integration testing shared between Rust and Python implementations:

- Single parameterized test suite in `tests/unit/mcpgateway/plugins/plugins/pii_filter/test_pii_filter.py`
- Tests run against both implementations using `@pytest.mark.parametrize`
- Graceful fallback when Rust plugins unavailable
- CI enforces Rust plugin availability via `REQUIRE_RUST_PLUGINS=1`

### CI Consolidation

- Removed separate `rust-plugins.yml` workflow
- Rust environment setup integrated into main `pytest.yml`
- All tests run through single parameterized pytest suite
- Eliminates duplicate test execution and maintenance burden

### This PR has 3 independent parts; found issues with:

License check was broken → added commit to fix it
Exfil plugin Rust version had bugs → needed to be addressed
Main plugin refactoring

### CODE REVIEW

File: .github/workflows/rust-plugins.yml [hunk 1]
```diff
@@ -53,30 +70,32 @@ jobs:
...
-          path: plugins_rust/target
-          key: ${{ runner.os }}-cargo-build-${{ hashFiles('**/Cargo.lock') }}
+          path: plugins_rust/${{ matrix.plugin }}/target
+          key: ${{ runner.os }}-${{ matrix.plugin }}-cargo-build-${{ hashFiles(format('plugins_rust/{0}/Cargo.lock', matrix.plugin)) }}
 
       - name: Check formatting
-        working-directory: plugins_rust
-        run: make fmt-check
+        working-directory: plugins_rust/${{ matrix.plugin }}
+        run: cargo fmt -- --check
```
Comment (brian-hussey) [inline .github/workflows/rust-plugins.yml]: Will you change these back to make targets that in turn call the appropriate cargo commands?
Make is the defacto standard for anything and the reference to see how any of these stages are executed for the people that don't necessarily know the different tech we have.
We still have the make targets but they run the risk of being splintered from pipeline if the GHA files have different targets to manage.

Comment (brian-hussey) [reply .github/workflows/rust-plugins.yml]: This applies to all the changes from make <target> to something else.

File: .github/workflows/rust-plugins.yml [hunk 2]
```diff
@@ -204,22 +210,43 @@ jobs:
...
+            maturin develop --release || exit 1
+            
+            if [ -f Makefile ] && grep -q "coverage:" Makefile; then
+              uv run make coverage || exit 1
+            else
+              cargo tarpaulin --out Xml --output-dir coverage || exit 1
+            fi
+            
+            cd ..
+          done
```
Comment (brian-hussey) [inline .github/workflows/rust-plugins.yml]: This should also be a single make target so that we can run it locally as well as via pipeline.

Comment (brian-hussey) [reply .github/workflows/rust-plugins.yml]: And any other targets like this should be contained as make targets instead of coded out here.

File: .github/workflows/rust-plugins.yml (hunk 1 above)
Comment (lucarlig) [reply .github/workflows/rust-plugins.yml]: this is @crivetimihai change, i had originally used make, what's the strategy make or raw command in CI?

File: .github/workflows/rust-plugins.yml (hunk 2 above)
Comment (lucarlig) [reply .github/workflows/rust-plugins.yml]: ok makes sense

**Comment by brian-hussey** [comment]
@crivetimihai I'm asking @lucarlig to change back to make targets so we have the ability to execute locally and through pipeline. Your commit on top of his work did the initial change for that. 
Was that intentional on your part or should we stick to make being the source of truth for tasks both locally and in pipeline stages?

**Comment by brian-hussey** [comment]
@lucarlig I'm also seeing issues with local execution targets:
following 
```
make venv install-dev
. .~/.venv/mcpgateway/bin/activate
pytest tests/unit/plugins/test_encoded_exfil_detector.py
```

without `export REQUIRE_RUST_PLUGINS=1`:
```
.s.s.s.s.s.s.s.s.s.s.s.....                                                                        [100%]
======================================== short test summary info =========================================
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:37: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:47: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:57: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:67: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:77: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:102: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:117: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:129: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:140: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:154: Rust not available
SKIPPED [1] tests/unit/plugins/test_encoded_exfil_detector.py:167: Rust not available
16 passed, 11 skipped in 0.32s
```

And with `export REQUIRE_RUST_PLUGINS=1`:
```

================================================= ERRORS =================================================
___________________ ERROR collecting tests/unit/plugins/test_encoded_exfil_detector.py ___________________
ImportError while importing test module '/Users/brian/dev/cf_review/pr/2961/tests/unit/plugins/test_encoded_exfil_detector.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/plugins/test_encoded_exfil_detector.py:17: in <module>
    import encoded_exfil_detection as _rust_encoded_exfil_detection  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   ModuleNotFoundError: No module named 'encoded_exfil_detection'

During handling of the above exception, another exception occurred:
../../../../.local/share/uv/python/cpython-3.13.9-macos-aarch64-none/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tests/unit/plugins/test_encoded_exfil_detector.py:24: in <module>
    raise ImportError("Rust plugin 'encoded_exfil_detection' is required in CI but not available")
E   ImportError: Rust plugin 'encoded_exfil_detection' is required in CI but not available
-------------------------------------------- Captured stderr ---------------------------------------------
2026-02-16T10:49:30 - mcpgateway.config - INFO - Using SQLite database. Consider PostgreSQL or MySQL for production.
2026-02-16T10:49:30 - mcpgateway.config - WARNING - 🔓 SECURITY WARNING: Default BASIC_AUTH_PASSWORD detected! Please change it if you enable API_ALLOW_BASIC_AUTH.
2026-02-16T10:49:30 - mcpgateway.config - WARNING - 🔐 SECURITY WARNING: Admin password has low complexity. Should contain at least 3 of: uppercase, lowercase, digits, special characters
2026-02-16T10:49:30 - mcpgateway.config - WARNING - 🔓 SECURITY WARNING - jwt_secret_key: Default/weak secret detected! Please set a strong, unique value for production.
2026-02-16T10:49:30 - mcpgateway.config - WARNING - ⚠️  SECURITY WARNING - jwt_secret_key: Secret should be at least 32 characters long. Current length: 11
2026-02-16T10:49:30 - mcpgateway.config - WARNING - 🔑 SECURITY WARNING - jwt_secret_key: Secret has low entropy. Consider using a more random value.
2026-02-16T10:49:30 - mcpgateway.config - WARNING - 🔓 SECURITY WARNING - auth_encryption_secret: Default/weak secret detected! Please set a strong, unique value for production.
2026-02-16T10:49:30 - mcpgateway.config - WARNING - ⚠️  SECURITY WARNING - auth_encryption_secret: Secret should be at least 32 characters long. Current length: 12
2026-02-16T10:49:30 - mcpgateway.config - WARNING - 🔑 SECURITY WARNING - auth_encryption_secret: Secret has low entropy. Consider using a more random value.
2026-02-16T10:49:30 - mcpgateway.db - INFO - Configuring SQLite with pool_size=50, max_overflow=10
======================================== short test summary info =========================================
ERROR tests/unit/plugins/test_encoded_exfil_detector.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
1 error in 0.34s
```

Other tests act this same way.
Can we add whatever dependencies are required to setting up the environment prior to anyone executing test. If we need to discuss the best way to handle this let me know.

**Comment by lucarlig** [comment]
@crivetimihai should be ready to merge, please review.

**Comment by cafalchio** [comment]
Thank you @lucarlig 
The install and tests are running well in a new venv. 
Should the rust install also export the REQUIRE_RUST_PLUGINS=1?

**Comment by lucarlig** [comment]
> Thank you @lucarlig The install and tests are running well in a new venv. Should the rust install also export the REQUIRE_RUST_PLUGINS=1?

i think as some point rust become a must and we remove this flag altogether, but for now should be ok, rust test can be skipped.

**Comment by brian-hussey** [comment]
We're getting a repeated error in the Rust Plugins CI/CD/Code Coverage:
```
error[E0004]: non-exhaustive patterns: `Some(Protocol::Tlsv13)` not covered
  --> /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/native-tls-0.2.17/src/imp/openssl.rs:61:22
   |
61 |     options |= match min {
   |                      ^^^ pattern `Some(Protocol::Tlsv13)` not covered
   |
note: `Option<Protocol>` defined here
  --> /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/option.rs:600:1
  ::: /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/option.rs:608:5
   |
   = note: not covered
   = note: the matched value is of type `Option<Protocol>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
   |
73 ~         },
74 +         Some(Protocol::Tlsv13) => todo!()
   |

error[E0004]: non-exhaustive patterns: `Some(Protocol::Tlsv13)` not covered
  --> /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/native-tls-0.2.17/src/imp/openssl.rs:75:22
   |
75 |     options |= match max {
   |                      ^^^ pattern `Some(Protocol::Tlsv13)` not covered
   |
note: `Option<Protocol>` defined here
  --> /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/option.rs:600:1
  ::: /rustc/254b59607d4417e9dffbc307138ae5c86280fe4c/library/core/src/option.rs:608:5
   |
   = note: not covered
   = note: the matched value is of type `Option<Protocol>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
   |
81 ~         },
82 +         Some(Protocol::Tlsv13) => todo!()
   |

For more information about this error, try `rustc --explain E0004`.
error: could not compile `native-tls` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: failed to compile `cargo-tarpaulin v0.35.1`, intermediate artifacts can be found at `/tmp/cargo-installT8H9TM`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
Error: Process completed with exit code 101.
```

Is this a problem with underlying dependencies of our code or of the build tools themselves?
Locally I'm getting dependencies resolved, and coverage reports generated, but nowhere near 100% coverage. 

@crivetimihai is that a blocker for the rust plugins at the moment or can we deal with that later?
I'm hesitant to think about joining the coverage reports for the whole project until we lock in closer to the full coverage for the overall quality gate.

**Comment by lucarlig** [comment]
> Is this a problem with underlying dependencies of our code or of the build tools themselves? Locally I'm getting dependencies resolved, and coverage reports generated, but nowhere near 100% coverage.
> 
> @crivetimihai is that a blocker for the rust plugins at the moment or can we deal with that later? I'm hesitant to think about joining the coverage reports for the whole project until we lock in closer to the full coverage for the overall quality gate.

I am working on fixing the coverage CI build issue (cargo tarpaulin installation seems broken), regarding the coverage % since the test are split in python and rust, it's a bit more complicated, I have already opened an issue for it https://github.com/IBM/mcp-context-forge/issues/2969

**Comment by lucarlig** [comment]
i switched out `cargo-tarpaulin` with `cargo-llvm-cov`. Seems better mantained.

**Review by crivetimihai — DISMISSED** [review]
Good architectural direction — moving from monolithic workspace to independent crates improves plugin isolation and scalability. However, several issues need addressing.

**Strengths:**
- Excellent Makefile design: auto-discovery via `find`, pattern rules for delegation, scales to N plugins automatically
- Type stub generation with pyo3-stub-gen improves IDE support
- Consistent crate configuration: `opt-level=3`, `lto="fat"`, `strip=true` for release profiles
- BSN (Dutch national ID) detection addition extends PII coverage
- `REQUIRE_RUST_PLUGINS=1` enforcement in CI
- Each plugin now independently auditable via `cargo audit`

**Issues requiring changes:**

1. **Python integration tests removed from CI**: The old `python-integration` job that ran `make test-python` and `make test-differential` was deleted from `rust-plugins.yml`. These differential tests (comparing Rust vs Python output) are critical for correctness. Verify these run elsewhere (e.g., in main pytest with `REQUIRE_RUST_PLUGINS=1`) or restore them.

2. **Import path breaking change without migration guide**: `from plugins_rust import PIIDetectorRust` → `from pii_filter_rust import PIIDetectorRust`. This breaks any external code using the old import. Need migration documentation.

3. **Wheel output path mismatch in Docker**: Containerfiles changed to copy from `/build/rust-wheels/` but the Makefile targets produce wheels in `plugins_rust/*/target/wheels/`. Verify the build stage actually populates `/build/rust-wheels/`.

4. **Crate naming inconsistency**: `pii_filter_rust` (with suffix), `secrets_detection_rust` (with suffix), but `encoded_exfil_detection` (without). Standardize naming.

5. **QUICKSTART.md deleted without replacement**: The quick-start guide for plugin developers was removed but no equivalent content was added to the updated README or docs. Plugin developers need onboarding documentation.

6. **No shared dependency management**: pyo3 version must now be managed in 3+ separate `Cargo.toml` files. No CI check enforces consistency. A security patch for pyo3 requires updating multiple files manually. Consider adding a CI lint that verifies key dependency versions are consistent across crates.

**Minor observations:**
- Benchmark tests only run on Ubuntu; should test macOS/Windows for architecture-specific regressions
- Committing `Cargo.lock` per-crate is correct for reproducible wheel builds
- The `lto="fat"` adds significant compile time; dev profiles should consider `lto=false`
- BSN pattern validation tests should be included to prevent false positives on 9-digit numbers

**Comment by lucarlig** [comment]
thank you for review @crivetimihai , all suggested changes makes sense except 1. The tests are now parametric from the standard pytest which makes them also differential, differential tests were merged into the main unit/integration tests for plugins. The main pytest is ran from `pytest.yml`

**Comment by crivetimihai** [comment]
Thanks for the updates and clarification, @lucarlig. Good to know the differential tests are now parametric in the main pytest — that addresses concern #1.

CI/Containerfile references look updated. One remaining observation:

**Crate naming still inconsistent** — `pii_filter_rust` and `secrets_detection_rust` have the `_rust` suffix, while `encoded_exfil_detection` does not. The documentation also shows `plugin_rust` as a generic import in some places. Not blocking, but worth standardizing for a consistent developer experience.

**Comment by lucarlig** [comment]
I have fixed the naming, I will do one more review tomorrow and then hopefully should be ready

File: plugins/encoded_exfil_detection/encoded_exfil_detector.py [hunk 3]
```diff
@@ -42,7 +42,7 @@
 
 # Try to import Rust-accelerated implementation
 try:
-    import encoded_exfil_detection
+    from encoded_exfil_detection_rust.encoded_exfil_detection_rust import py_scan_container as encoded_exfil_detection
```
Comment (gandhipratik203) [inline plugins/encoded_exfil_detection/encoded_exfil_detector.py]: **Potential issue to verify:** Rust fast-path invocation may not be working as intended due to an import/call mismatch.
**Note**: this was an AI review suggestion. But, if the impact/execution flow described above is accurate, it may be worth a quick validation and diagnosis.

In this file, `py_scan_container` is imported as a function alias:

`from encoded_exfil_detection_rust.encoded_exfil_detection_rust import py_scan_container as encoded_exfil_detection`

So `encoded_exfil_detection` is a callable function.  
But later in `_scan_container`, it is invoked as if it were a module/object:

`encoded_exfil_detection.py_scan_container(container, cfg)`

That can raise at runtime (`function` has no attribute `py_scan_container`), and because this block is wrapped in `try/except`, execution falls back to the Python implementation.

Impact:
- Rust acceleration may not execute for this plugin.
- Behavior can still appear correct because fallback masks the issue.
- Performance and Rust-path coverage may be lower than expected without obvious signals.

Suggested fix: use one consistent pattern:
1) function import + direct call:
   `count, redacted, findings = encoded_exfil_detection(container, cfg)`
or
2) module import + attribute call:
   `import encoded_exfil_detection_rust.encoded_exfil_detection_rust as encoded_exfil_detection`
   then keep `.py_scan_container(...)`.


If this diagnosis is accurate, a similar import/call pattern may also be worth validating in plugins/secrets_detection/secrets_detection.py.

**Comment by lucarlig** [comment]
@crivetimihai ready, please review

**Review by cafalchio — DISMISSED** [review]
Hi Luca, thanks for organising a bit more the rust plugins. LGTM

**Review by gandhipratik203 — DISMISSED** [review]
Looks good overall, and the restructure should help other plugins. Approving.

**Comment by crivetimihai** [comment]
## Review Summary

Rebased onto main (clean, no conflicts) and performed a thorough review of the architecture, Rust code, Python wrappers, CI/build system, tests, and security posture.

### Architecture
Sound decision to migrate from Cargo workspace to independent plugin crates. Each plugin now has its own `Cargo.toml`, `pyproject.toml`, `Makefile`, and `Cargo.lock` — enabling independent build, test, versioning, and PyPI distribution. The double-nested import pattern (`from pii_filter_rust.pii_filter_rust import PIIDetectorRust`) is correctly documented and consistent across all three plugins.

### Bug Found & Fixed
**`encoded_exfil_detector.py:314`** — Called `encoded_exfil_detection.py_scan_container()` as an attribute access, but `encoded_exfil_detection` was already aliased to the `py_scan_container` function via the import. This would raise `AttributeError` at runtime when Rust is available, silently falling back to Python. Fixed to call the function directly: `encoded_exfil_detection(container, cfg)`.

### Rust Code
- No unsafe code, proper bounds checking, safe PyO3 bindings
- `LazyLock` for regex initialization, `Cow` for zero-copy where applicable
- Overlap detection, reverse-order redaction, and exhaustive enum handling all correct
- Minor optimization opportunity: `has_egress_context()` allocates via `.to_lowercase()` per-candidate (acceptable for security tooling)

### Tests
- Old differential tests (788 lines) and Rust-specific tests (533 lines) properly consolidated into parameterized suites
- Added 15 new unit tests for `encoded_exfil_detector.py` helper functions covering edge cases (empty data, padding, decode failures, oversized input, disabled encodings, container dispatch, finding limits)
- All 138 Python tests pass; Rust-only tests properly skipped when Rust unavailable
- Differential coverage: 95% on the new `encoded_exfil_detector.py` (remaining 5% is Rust-only paths)

### Security
- No hardcoded secrets, no command injection vectors in Makefiles
- Containerfiles use multi-stage builds, non-root user, HTTPS-only for rustup
- Plugin import fallbacks are properly guarded
- Pre-commit exclusions for Rust test patterns are appropriate

**Comment by crivetimihai** [comment]
Reopened as #3147. CI/CD will re-run on the new PR. You are still credited as the author.
